File: tests/test_globals.cpp
/**
 * Test: test_globals
 * Tests LLVM global variable operations
 *
 * LLVM-C APIs covered:
 * - LLVMAddGlobal(), LLVMAddGlobalInAddressSpace()
 * - LLVMGetNamedGlobal()
 * - LLVMSetInitializer(), LLVMGetInitializer()
 * - LLVMSetGlobalConstant(), LLVMIsGlobalConstant()
 * - LLVMSetLinkage(), LLVMGetLinkage()
 * - LLVMSetVisibility(), LLVMGetVisibility()
 * - LLVMSetAlignment(), LLVMGetAlignment()
 * - LLVMSetSection(), LLVMGetSection()
 * - LLVMGetFirstGlobal(), LLVMGetNextGlobal(), LLVMGetLastGlobal()
 * - LLVMDeleteGlobal()
 * - LLVMSetThreadLocal(), LLVMIsThreadLocal()
 * - LLVMSetExternallyInitialized(), LLVMIsExternallyInitialized()
 */

#include <cstdio>
#include <llvm-c/Analysis.h>
#include <llvm-c/Core.h>

const char *linkage_name(LLVMLinkage linkage) {
  switch (linkage) {
  case LLVMExternalLinkage:
    return "external";
  case LLVMAvailableExternallyLinkage:
    return "available_externally";
  case LLVMLinkOnceAnyLinkage:
    return "linkonce";
  case LLVMLinkOnceODRLinkage:
    return "linkonce_odr";
  case LLVMWeakAnyLinkage:
    return "weak";
  case LLVMWeakODRLinkage:
    return "weak_odr";
  case LLVMAppendingLinkage:
    return "appending";
  case LLVMInternalLinkage:
    return "internal";
  case LLVMPrivateLinkage:
    return "private";
  case LLVMExternalWeakLinkage:
    return "extern_weak";
  case LLVMCommonLinkage:
    return "common";
  default:
    return "unknown";
  }
}

const char *visibility_name(LLVMVisibility vis) {
  switch (vis) {
  case LLVMDefaultVisibility:
    return "default";
  case LLVMHiddenVisibility:
    return "hidden";
  case LLVMProtectedVisibility:
    return "protected";
  default:
    return "unknown";
  }
}

int main() {
  LLVMContextRef ctx = LLVMContextCreate();
  LLVMModuleRef mod = LLVMModuleCreateWithNameInContext("test_globals", ctx);

  LLVMTypeRef i8 = LLVMInt8TypeInContext(ctx);
  LLVMTypeRef i32 = LLVMInt32TypeInContext(ctx);
  LLVMTypeRef i64 = LLVMInt64TypeInContext(ctx);
  LLVMTypeRef f64 = LLVMDoubleTypeInContext(ctx);
  LLVMTypeRef ptr = LLVMPointerTypeInContext(ctx, 0);

  // ==========================================
  // Basic global variable
  // ==========================================
  LLVMValueRef global_counter = LLVMAddGlobal(mod, i32, "counter");
  LLVMSetInitializer(global_counter, LLVMConstInt(i32, 0, 0));

  // ==========================================
  // Constant global
  // ==========================================
  LLVMValueRef global_const = LLVMAddGlobal(mod, i32, "magic_number");
  LLVMSetInitializer(global_const, LLVMConstInt(i32, 42, 0));
  LLVMSetGlobalConstant(global_const, 1);

  // ==========================================
  // Global with alignment
  // ==========================================
  LLVMValueRef global_aligned = LLVMAddGlobal(mod, i64, "aligned_var");
  LLVMSetInitializer(global_aligned, LLVMConstInt(i64, 0, 0));
  LLVMSetAlignment(global_aligned, 16);

  // ==========================================
  // Global with linkage
  // ==========================================
  LLVMValueRef global_internal = LLVMAddGlobal(mod, i32, "internal_var");
  LLVMSetInitializer(global_internal, LLVMConstInt(i32, 100, 0));
  LLVMSetLinkage(global_internal, LLVMInternalLinkage);

  LLVMValueRef global_private = LLVMAddGlobal(mod, i32, "private_var");
  LLVMSetInitializer(global_private, LLVMConstInt(i32, 200, 0));
  LLVMSetLinkage(global_private, LLVMPrivateLinkage);

  LLVMValueRef global_weak = LLVMAddGlobal(mod, i32, "weak_var");
  LLVMSetInitializer(global_weak, LLVMConstInt(i32, 300, 0));
  LLVMSetLinkage(global_weak, LLVMWeakAnyLinkage);

  // ==========================================
  // Global with visibility
  // ==========================================
  LLVMValueRef global_hidden = LLVMAddGlobal(mod, i32, "hidden_var");
  LLVMSetInitializer(global_hidden, LLVMConstInt(i32, 0, 0));
  LLVMSetVisibility(global_hidden, LLVMHiddenVisibility);

  // ==========================================
  // Global with section
  // ==========================================
  LLVMValueRef global_section = LLVMAddGlobal(mod, i32, "section_var");
  LLVMSetInitializer(global_section, LLVMConstInt(i32, 0, 0));
  LLVMSetSection(global_section, ".mydata");

  // ==========================================
  // Thread-local global
  // ==========================================
  LLVMValueRef global_tls = LLVMAddGlobal(mod, i32, "tls_var");
  LLVMSetInitializer(global_tls, LLVMConstInt(i32, 0, 0));
  LLVMSetThreadLocal(global_tls, 1);

  // ==========================================
  // Externally initialized global (no initializer)
  // ==========================================
  LLVMValueRef global_extern = LLVMAddGlobal(mod, i32, "extern_var");
  LLVMSetExternallyInitialized(global_extern, 1);

  // ==========================================
  // Global in address space
  // ==========================================
  LLVMValueRef global_addrspace =
      LLVMAddGlobalInAddressSpace(mod, i32, "addrspace_var", 1);
  LLVMSetInitializer(global_addrspace, LLVMConstInt(i32, 0, 0));

  // ==========================================
  // Global to be deleted
  // ==========================================
  LLVMValueRef global_delete = LLVMAddGlobal(mod, i32, "to_be_deleted");
  LLVMSetInitializer(global_delete, LLVMConstInt(i32, 999, 0));

  // Count globals before deletion
  int count_before = 0;
  for (LLVMValueRef g = LLVMGetFirstGlobal(mod); g; g = LLVMGetNextGlobal(g)) {
    count_before++;
  }

  // Delete the global
  LLVMDeleteGlobal(global_delete);

  // Count globals after deletion
  int count_after = 0;
  for (LLVMValueRef g = LLVMGetFirstGlobal(mod); g; g = LLVMGetNextGlobal(g)) {
    count_after++;
  }

  // Get global by name
  LLVMValueRef found_counter = LLVMGetNamedGlobal(mod, "counter");
  LLVMValueRef found_nonexist = LLVMGetNamedGlobal(mod, "nonexistent");

  // Get initializer
  LLVMValueRef init = LLVMGetInitializer(global_const);

  // Verify module
  char *error = nullptr;
  if (LLVMVerifyModule(mod, LLVMReturnStatusAction, &error)) {
    fprintf(stderr, "; Verification failed: %s\n", error);
    LLVMDisposeMessage(error);
    LLVMDisposeModule(mod);
    LLVMContextDispose(ctx);
    return 1;
  }
  LLVMDisposeMessage(error);

  // Print diagnostic comments
  printf("; Test: test_globals\n");
  printf(";\n");
  printf("; Global variable properties:\n");
  printf(";\n");
  printf("; counter:\n");
  printf(";   is constant: %s\n",
         LLVMIsGlobalConstant(global_counter) ? "yes" : "no");
  printf(";   linkage: %s\n", linkage_name(LLVMGetLinkage(global_counter)));
  printf(";\n");
  printf("; magic_number:\n");
  printf(";   is constant: %s\n",
         LLVMIsGlobalConstant(global_const) ? "yes" : "no");
  printf(";   has initializer: %s\n", init != nullptr ? "yes" : "no");
  printf(";   initializer value: %llu\n", LLVMConstIntGetZExtValue(init));
  printf(";\n");
  printf("; aligned_var:\n");
  printf(";   alignment: %u\n", LLVMGetAlignment(global_aligned));
  printf(";\n");
  printf("; internal_var:\n");
  printf(";   linkage: %s\n", linkage_name(LLVMGetLinkage(global_internal)));
  printf(";\n");
  printf("; hidden_var:\n");
  printf(";   visibility: %s\n",
         visibility_name(LLVMGetVisibility(global_hidden)));
  printf(";\n");
  printf("; section_var:\n");
  printf(";   section: %s\n", LLVMGetSection(global_section));
  printf(";\n");
  printf("; tls_var:\n");
  printf(";   is thread local: %s\n",
         LLVMIsThreadLocal(global_tls) ? "yes" : "no");
  printf(";\n");
  printf("; extern_var:\n");
  printf(";   is externally initialized: %s\n",
         LLVMIsExternallyInitialized(global_extern) ? "yes" : "no");
  printf(";\n");
  printf("; Lookup tests:\n");
  printf(";   found 'counter': %s\n", found_counter != nullptr ? "yes" : "no");
  printf(";   found 'nonexistent': %s\n",
         found_nonexist != nullptr ? "yes" : "no");
  printf(";\n");
  printf("; Global counts:\n");
  printf(";   before deletion: %d\n", count_before);
  printf(";   after deletion: %d\n", count_after);
  printf(";\n");
  printf("; All globals:\n");
  for (LLVMValueRef g = LLVMGetFirstGlobal(mod); g; g = LLVMGetNextGlobal(g)) {
    size_t len;
    const char *name = LLVMGetValueName2(g, &len);
    printf(";   - %s\n", name);
  }
  printf("\n");

  // Print module IR
  char *ir = LLVMPrintModuleToString(mod);
  printf("%s", ir);
  LLVMDisposeMessage(ir);

  // Cleanup
  LLVMDisposeModule(mod);
  LLVMContextDispose(ctx);

  return 0;
}

File: tests/test_globals.py
#!/usr/bin/env -S uv run
"""
Test: test_globals
Tests LLVM global variable operations

Python equivalent of tests/test_globals.cpp
Must produce identical output to the C++ version.
"""

import llvm


def linkage_name(linkage):
    names = {
        llvm.Linkage.External: "external",
        llvm.Linkage.AvailableExternally: "available_externally",
        llvm.Linkage.LinkOnceAny: "linkonce",
        llvm.Linkage.LinkOnceODR: "linkonce_odr",
        llvm.Linkage.WeakAny: "weak",
        llvm.Linkage.WeakODR: "weak_odr",
        llvm.Linkage.Appending: "appending",
        llvm.Linkage.Internal: "internal",
        llvm.Linkage.Private: "private",
        llvm.Linkage.ExternalWeak: "extern_weak",
        llvm.Linkage.Common: "common",
    }
    return names.get(linkage, "unknown")


def visibility_name(vis):
    names = {
        llvm.Visibility.Default: "default",
        llvm.Visibility.Hidden: "hidden",
        llvm.Visibility.Protected: "protected",
    }
    return names.get(vis, "unknown")


def main():
    with llvm.create_context() as ctx:
        with ctx.create_module("test_globals") as mod:
            i8 = ctx.types.i8
            i32 = ctx.types.i32
            i64 = ctx.types.i64
            f64 = ctx.types.f64
            ptr = ctx.types.ptr

            # ==========================================
            # Basic global variable
            # ==========================================
            global_counter = mod.add_global(i32, "counter")
            global_counter.initializer = i32.constant(0)

            # ==========================================
            # Constant global
            # ==========================================
            global_const = mod.add_global(i32, "magic_number")
            global_const.initializer = i32.constant(42)
            global_const.set_constant(True)

            # ==========================================
            # Global with alignment
            # ==========================================
            global_aligned = mod.add_global(i64, "aligned_var")
            global_aligned.initializer = i64.constant(0)
            global_aligned.alignment = 16

            # ==========================================
            # Global with linkage
            # ==========================================
            global_internal = mod.add_global(i32, "internal_var")
            global_internal.initializer = i32.constant(100)
            global_internal.linkage = llvm.Linkage.Internal

            global_private = mod.add_global(i32, "private_var")
            global_private.initializer = i32.constant(200)
            global_private.linkage = llvm.Linkage.Private

            global_weak = mod.add_global(i32, "weak_var")
            global_weak.initializer = i32.constant(300)
            global_weak.linkage = llvm.Linkage.WeakAny

            # ==========================================
            # Global with visibility
            # ==========================================
            global_hidden = mod.add_global(i32, "hidden_var")
            global_hidden.initializer = i32.constant(0)
            global_hidden.visibility = llvm.Visibility.Hidden

            # ==========================================
            # Global with section
            # ==========================================
            global_section = mod.add_global(i32, "section_var")
            global_section.initializer = i32.constant(0)
            global_section.section = ".mydata"

            # ==========================================
            # Thread-local global
            # ==========================================
            global_tls = mod.add_global(i32, "tls_var")
            global_tls.initializer = i32.constant(0)
            global_tls.set_thread_local(True)

            # ==========================================
            # Externally initialized global (no initializer)
            # ==========================================
            global_extern = mod.add_global(i32, "extern_var")
            global_extern.set_externally_initialized(True)

            # ==========================================
            # Global in address space
            # ==========================================
            global_addrspace = mod.add_global_in_address_space(i32, "addrspace_var", 1)
            global_addrspace.initializer = i32.constant(0)

            # ==========================================
            # Global to be deleted
            # ==========================================
            global_delete = mod.add_global(i32, "to_be_deleted")
            global_delete.initializer = i32.constant(999)

            # Count globals before deletion
            count_before = len(mod.globals)

            # Delete the global
            global_delete.delete_global()

            # Count globals after deletion
            count_after = len(mod.globals)

            # Get global by name
            found_counter = mod.get_global("counter")
            found_nonexist = mod.get_global("nonexistent")

            # Get initializer
            init = global_const.initializer

            # Verify module
            if not mod.verify():
                print(f"; Verification failed: {mod.get_verification_error()}")
                return 1

            # Print diagnostic comments
            print("; Test: test_globals")
            print(";")
            print("; Global variable properties:")
            print(";")
            print("; counter:")
            print(
                f";   is constant: {'yes' if global_counter.is_global_constant else 'no'}"
            )
            print(f";   linkage: {linkage_name(global_counter.linkage)}")
            print(";")
            print("; magic_number:")
            print(
                f";   is constant: {'yes' if global_const.is_global_constant else 'no'}"
            )
            print(f";   has initializer: {'yes' if init else 'no'}")
            if init:
                print(f";   initializer value: {init.const_zext_value}")
            else:
                print(";   initializer value: None")
            print(";")
            print("; aligned_var:")
            print(f";   alignment: {global_aligned.alignment}")
            print(";")
            print("; internal_var:")
            print(f";   linkage: {linkage_name(global_internal.linkage)}")
            print(";")
            print("; hidden_var:")
            print(f";   visibility: {visibility_name(global_hidden.visibility)}")
            print(";")
            print("; section_var:")
            print(f";   section: {global_section.section}")
            print(";")
            print("; tls_var:")
            print(
                f";   is thread local: {'yes' if global_tls.is_thread_local else 'no'}"
            )
            print(";")
            print("; extern_var:")
            print(
                f";   is externally initialized: {'yes' if global_extern.is_externally_initialized else 'no'}"
            )
            print(";")
            print("; Lookup tests:")
            print(f";   found 'counter': {'yes' if found_counter else 'no'}")
            print(f";   found 'nonexistent': {'yes' if found_nonexist else 'no'}")
            print(";")
            print("; Global counts:")
            print(f";   before deletion: {count_before}")
            print(f";   after deletion: {count_after}")
            print(";")
            print("; All globals:")
            for g in mod.globals:
                print(f";   - {g.name}")
            print()

            # Print module IR
            print(mod.to_string(), end="")

    return 0


if __name__ == "__main__":
    exit(main())
